File: replication/CURRENT_SOURCE_EVIDENCE/V4_1_IMPLEMENTATION_A/bin/classify_v4_1.py
#!/usr/bin/env python3
"""Command-line entry point for one exact archive."""

import argparse
import json
import sys
from pathlib import Path


IMPLEMENTATION_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(IMPLEMENTATION_ROOT))

from v4a.classifier import classify_archive  # noqa: E402


def canonical_json_bytes(value):
    return (json.dumps(value, ensure_ascii=False, sort_keys=True, separators=(",", ":")) + "\n").encode(
        "utf-8"
    )


def parse_args():
    parser = argparse.ArgumentParser(description="Classify one exact V4 source archive without extraction")
    parser.add_argument("--archive", required=True, type=Path)
    parser.add_argument("--audit-id", required=True)
    parser.add_argument("--expected-sha256", required=True)
    parser.add_argument("--expected-bytes", required=True, type=int)
    parser.add_argument("--registry", required=True, type=Path)
    return parser.parse_args()


def main():
    args = parse_args()
    registry = json.loads(args.registry.read_text(encoding="utf-8"))
    result = classify_archive(
        archive_path=args.archive,
        audit_id=args.audit_id,
        expected_sha256=args.expected_sha256,
        expected_bytes=args.expected_bytes,
        registry=registry,
    )
    sys.stdout.buffer.write(canonical_json_bytes(result))


if __name__ == "__main__":
    main()

File: replication/CURRENT_SOURCE_EVIDENCE/V4_1_IMPLEMENTATION_B/bin/classify_v4_1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "json"
require "optparse"
require_relative "../lib/reference_v4_b"

options = {}
OptionParser.new do |parser|
  parser.on("--archive PATH") { |value| options[:archive] = value }
  parser.on("--audit-id ID") { |value| options[:audit_id] = value }
  parser.on("--expected-sha256 SHA") { |value| options[:expected_sha256] = value }
  parser.on("--expected-bytes N", Integer) { |value| options[:expected_bytes] = value }
  parser.on("--registry PATH") { |value| options[:registry] = value }
end.parse!

required = %i[archive audit_id expected_sha256 expected_bytes registry]
missing = required.reject { |key| options.key?(key) }
abort("missing options: #{missing.join(', ')}") unless missing.empty?

registry = JSON.parse(File.read(options.fetch(:registry)))
classifier = ReferenceV4B::Classifier.new(registry)
result = classifier.classify(
  archive_path: options.fetch(:archive),
  audit_id: options.fetch(:audit_id),
  expected_sha256: options.fetch(:expected_sha256),
  expected_bytes: options.fetch(:expected_bytes)
)
STDOUT.write(JSON.generate(result))
STDOUT.write("\n")
